internal/gcs/bridge.go

```diff
@@ -0,0 +1,356 @@
+package gcs
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/Microsoft/hcsshim/internal/guid"
+	"github.com/sirupsen/logrus"
+)
+
+type requestMessage interface {
+	SetActivityID(id guid.GUID)
+}
+
+type responseMessage interface {
+	Err() error
+}
+
+// rpc represents an outstanding rpc request to the guest
+type rpc struct {
+	proc rpcProc
+	id   int64
+	req  requestMessage
+	resp responseMessage
+	err  error
+	ch   chan struct{}
+}
+
+// bridge represents a communcations bridge with the guest. It handles the
+// transport layer but (mostly) does not parse or construct the message payload.
+type bridge struct {
+	Timeout time.Duration
+	mu      sync.Mutex
+	nextID  int64
+	rpcs    map[int64]*rpc
+	conn    io.ReadWriteCloser
+	rpcCh   chan *rpc
+	notify  notifyFunc
+	closed  bool
+	log     *logrus.Entry
+}
+
+var (
+	errBridgeClosed = errors.New("bridge closed")
+
+	// Any non-async RPC must respond in this time or the bridge
+	// connection will be terminated.
+	bridgeFailureTimeout = time.Minute * 5
+)
+
+type notifyFunc func(*containerNotification) error
+
+// newBridge returns a bridge on the given connection. It calls notify when a
+// notification message arrives from the guest. It logs transport errors and
+// traces using the given logrus entry.
+func newBridge(conn io.ReadWriteCloser, notify notifyFunc, log *logrus.Entry) *bridge {
```

Comment:
FYI: Not asking you to change but with intellisense I have seen messages like this:

```go
// newBridge returns a bridge connected to `conn`. It calls `notify` when a ...
```

And then `conn` and `notify` which are the param names have syntax highlighting.